vike/shared/page-configs/Config.ts

```diff
@@ -39,6 +65,153 @@ type Config = ConfigBuiltIn &
+ */
+type OnHydrationEndAsync = (pageContext: PageContextClient) => Promise<void>
+/** Hook called after the page is hydrated.
+ *
+ * https://vike.dev/clientRouting
+ */
+type OnHydrationEndSync = (pageContext: PageContextClient) => void
+/** Hook called after the user navigates to a new page.
+ *
+ * https://vike.dev/clientRouting
+ */
+type OnPageTransitionEndAsync = (pageContext: PageContextClient) => Promise<void>
+/** Hook called after the user navigates to a new page.
+ *
+ * https://vike.dev/clientRouting
+ */
+type OnPageTransitionEndSync = (pageContext: PageContextClient) => void
+/** Hook called before the user navigates to a new page.
+ *
+ * https://vike.dev/clientRouting
+ */
+type OnPageTransitionStartAsync = (pageContext: PageContextClient) => Promise<void>
+/** Hook called before the user navigates to a new page.
+ *
+ * https://vike.dev/clientRouting
+ */
+type OnPageTransitionStartSync = (pageContext: PageContextClient) => void
+/** Page Hook called when pre-rendering starts.
+ *
+ * https://vike.dev/onPrerenderStart
+ */
+type OnPrerenderStartAsync = (prerenderContext: {
+  pageContexts: PageContextServer[]
+}) => Promise<{ prerenderContext: { pageContexts: PageContextServer[] } }>
+/** Page Hook called when pre-rendering starts.
+ *
+ * https://vike.dev/onPrerenderStart
+ */
+type OnPrerenderStartSync = (prerenderContext: { pageContexts: PageContextServer[] }) => {
+  prerenderContext: { pageContexts: PageContextServer[] }
+}
+/** Hook called when page is rendered on the client-side.
+ *
+ * https://vike.dev/onRenderClient
+ */
+type OnRenderClientAsync = (pageContext: PageContextClient) => Promise<void>
+/** Hook called when page is rendered on the client-side.
+ *
+ * https://vike.dev/onRenderClient
+ */
+type OnRenderClientSync = (pageContext: PageContextClient) => void
+/** Hook called when page is rendered to HTML on the server-side.
+ *
+ * https://vike.dev/onRenderHtml
+ */
+type OnRenderHtmlAsync = (pageContext: PageContextServer) => Promise<
+  | DocumentHtml
+  | {
+      documentHtml: DocumentHtml
+      // See https://vike.dev/stream#initial-data-after-stream-end
+      pageContext: Partial<Vike.PageContext> | (() => Promise<Partial<Vike.PageContext>>)
+    }
+>
+/** Hook called when page is rendered to HTML on the server-side.
+ *
+ * https://vike.dev/onRenderHtml
+ */
+type OnRenderHtmlSync = (pageContext: PageContextServer) =>
+  | DocumentHtml
+  | {
+      documentHtml: DocumentHtml
+      // See https://vike.dev/stream#initial-data-after-stream-end
+      pageContext: Partial<Vike.PageContext> | (() => Promise<Partial<Vike.PageContext>>)
+    }
+/** @deprecated Use a sync route() with an async guard() instead */
+type RouteAsync = (pageContext: PageContextServer | PageContextClient) => Promise<{ routeParams: Record<string, string> }>
+/** The page's URL(s).
+ *
+ *  https://vike.dev/route
+ */
+type RouteSync = (pageContext: PageContextServer | PageContextClient) => { routeParams: Record<string, string> }
```

Comment:
This won't work unfortunately, we had this topic earlier in this PR: `PageContextServer` won't be assignable to `PageContextServer | PageContextClient`. It works the other way around. In other words:

* `(A | B) => void` is assignable to `(A) => void`, but
* `(A) => void` is not assignable to `(A | B) => void`

So the user won't be able to stop using client routing and create a `route()` that takes only `PageContextServer` as argument. I can create a TS playground to illustrate if needed.

Reply:
Yes I know. My idea is to tell Server Routing users to not use the `Route` type. We can increasingly assume users to use Client Routing because of `vike-{react,vue,solid}`.

My thinking is: users who use `vike-{react,vue,solid}` are more likely to be junior-ish and it's particularly important for them to provide friendly types.

Current users who use Server Routing are "expert users"; I think it's fine to be less friendly. (Alternatively we can provide `RouteWithServerRouting`, like `PageContextClientWithServerRouting`.)

In a nutshell: we highly polish things for Client Routing and neglect Server Routing. (Honestly, I'm even considering deprecating Server Routing.)

Reply:
@brillout even an expert won't be able to perform such an assignment if they do it in `+config.h.ts`, similarly to the problem I was showing here: https://github.com/vikejs/vike/pull/1137#discussion_r1356928529

Reply:
In otherwords, yes if you use a `+route.ts` file then you are free not to use the `Route` type and no assignment type checking will be performed, true, but this won't work when directly assigning in `+config.h.ts`

Reply:
True 👍. A solution would be to use `pageContext: PageContextServer` for `Config['route']` (instead of setting it to `RouteSync | RouteAsync`). Alternatively, we create `RouteWithServerRouting` and set `Config['route']` to `RouteSync | RouteAsync | RouteWithServerRoutingSync | RouteWithServerRoutingAsync`. Either way is fine with me.

I know it's very ugly for Server Routing users and only slightly prettier for Client Routing users. That's how much I think we should prioritize Client Routing over Server Routing... (I'd be up for deprecating Server Routing after all).

Reply:
> A solution would be to use `pageContext: PageContextServer` for `Config['route']`

That's brilliant, let me push that

Reply:
actually, I think it's not necessary. As a user, if I disable Client Routing, and I'm forced to still define a `route(PageContextServer | PageContextClient)`, although I know `PageContextClient` will never be passed, so what? Why would this bother me? I'm still going to be able to write a pretty decent implementation. Am I missing something?

Reply:
(if you agree, let's merge :rocket:)

Reply:
👍 I think you're right, the slight mismatch won't bother Server Routing users that much.

To answer your question: yes, there are slight mismatches between `PageContextServer` and `PageContextClient`, the most notable one is that everything is `Partial` because TypeScript cannot now the value of `passToClient`. (See `vike/shared/types.ts`.)

Actually... I'm just realizing that `PageContextClient` is only correct for Client Routing users... For Server Routing users it should actually be `PageContextClientWithServerRouting`. I've just checked and `PageContextBuiltInClientWithServerRouting` and `PageContextBuiltInClientWithClientRouting` (that's the core diff) are quite similar though: I think this mismatch is ok.

This is actually a good example for why I'm inclined to deprecate Server Routing: in my experience the maintaince costs it adds isn't worth it.

After thinking more about it: I suggest we keep the implementation as is while we tell Server Routing users to not use these new types introduced by this PR. This clearly marks Server Routing users as second-class citizen which fits the long-term plan of encouraging users to use `vike-{react,vue,solid}` while eventually deprecating Server Routing as soon it becomes too annoying to maintain.

The issues we're having with it in this PR makes me think it's a reason enough to deprecate it. I'm now inclined to think we should deprecate Server Routing before releasing `v1.0.0`.

> (if you agree, let's merge )

👍 Let me know if you don't object with that last comment and I'll merge.

Reply:
This sounds scary - aren't there valid use cases for not wanting client routing? I'll trust your guts on this

Reply:
(whether or not we deprecate server routing, I think this PR is good enough to be merged)

Reply:
While Server Routing is marginally faster on first page load (less client-side code to load), Client Routing is tremendously faster on subsequent navigation. I find this is a reason enough to recommend Client Routing over Server Routing.

Also there are use cases that only work with Client Routing (e.g. a music player that should still play music when navigating pages). That's the biggest reason why I'd always recommend using Client Routing. (You never know what use case you may need in the future.)

The idea of supporting Server Routing was neat but, in hindsight, I think it was a mistake.

All in all, I think it's worth it to deprecate it.